lib/Runtime/Library/JavascriptGeneratorFunction.cpp

```diff
@@ -65,7 +65,7 @@ using namespace Js;
         JS_ETW(EventWriteJSCRIPT_RECYCLER_ALLOCATE_FUNCTION(scriptFunction, EtwTrace::GetFunctionId(functionProxy)));
 
         JavascriptGeneratorFunction* genFunc = nullptr;
-        if (functionProxy->IsAsync())
+        if (functionProxy->IsAsync() && !functionProxy->IsModule())
```

Comment:
This Function OP_NewScGenFunc is called when JS:
1. calls a generator function
2. calls an async function
3. calls an async generator function OR
4. begins evaluation of a module (i.e. the call from line 1046 of SourceTextModuleRecord.cpp)

Perhaps it would be better to split the command but for now in all 4 of those cases you end up here then have to determine the correct actions to take.

For modules the logic we need is in `JavascriptGeneratorFunction::New` but if the module contains a Top Level await it will be marked as an async function as well as a module - had to specifically check for modules to avoid calling `JavascriptAsyncFunction::New`.

The ToDo is a different point and is from a long time ago - the test case "Root and dynamic" appears to show that it works - though I am wondering if there's any more complex tests we can/should add.

Reply:
Makes sense, thank you!